src/main/scala/cromiam/samiam/SamIamPerAuthRequestActor.scala

```diff
@@ -0,0 +1,61 @@
+package cromiam.samiam
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props, Terminated}
+import cromiam.samiam.SamIamActor.{AuthorizationRequest, SamFailed}
+import cromiam.samiam.SamIamPerAuthRequestActor.{SamIamDenialException, SendMe}
+import cromiam.samiam.SamSupervisor.SamActorDown
+
+import scala.concurrent.Promise
+
+class SamIamPerAuthRequestActor(authorizationRequest: AuthorizationRequest,
+                                samIamActor: ActorRef,
+                                completeMe: Promise[Unit]) extends Actor with ActorLogging {
+
+  context.watch(samIamActor)
+  self ! SendMe(SamIamActor.RequestAuthorization(authorizationRequest))
+
+  override def receive: Receive = {
+    case SendMe(msg) => samIamActor ! msg
+    case authorized @ SamIamActor.UserActionAuthorized(ar) =>
+      if (ar == authorizationRequest) {
+        complete(())
+      } else {
+        complete(new Exception(s"Authorization response didn't correlate to request: SamIamActor.RequestAuthorization($ar) returned $authorized"))
+      }
+    case denied @ SamIamActor.UserActionDenied(ar) =>
+      if (ar == authorizationRequest) {
+        log.warning(s"Authorization denied by SamIAM: $ar")
+        complete(SamIamDenialException)
+      } else {
```

Comment:
if you had a common trait for these `UserAction...` things you could handle the unhappy cases in one spot.